org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/MessageJsonHandler.java

```diff
@@ -74,19 +103,39 @@ public void setMethodProvider(MethodProvider methodProvider) {
 		this.methodProvider = methodProvider;
 	}
 	
-	public Message parseMessage(CharSequence input) {
-		return gson.fromJson(input.toString(), Message.class);
+	public Message parseMessage(CharSequence input) throws JsonParseException {
+		StringReader reader = new StringReader(input.toString());
+		return parseMessage(reader);
 	}
 	
-	public Message parseMessage(Reader input) {
-		return gson.fromJson(input, Message.class);
+	public Message parseMessage(Reader input) throws JsonParseException {
+		JsonReader jsonReader = new JsonReader(input);
+		Message message = gson.fromJson(jsonReader, Message.class);
+		
+		if (message != null) {
+			// Check whether the input has been fully consumed
+			try {
+				if (jsonReader.peek() != JsonToken.END_DOCUMENT) {
+					MessageIssue issue = new MessageIssue("JSON document was not fully consumed.", ResponseErrorCode.ParseError.getValue());
+					throw new MessageIssueException(message, issue);
+				}
+			} catch (MalformedJsonException e) {
```

Comment:
Having a MessageIssue doesn’t seem necessary, you could just pass the exception to an MessageExceptionHandler instead. Now you’re catching exceptions, throwing away a bunch of information like the stack trace, then passing that object on.

Reply:
The exception and its stack trace are still available in the MessageIssue (`getCause()`).

The MessageJsonHandler and the MessageTypeAdapter make sure that malformed JSON code is not just logged locally, but forwarded to the remote endpoint in the form of a ResponseMessage.